File: lab2/LearnerRefal/mat/mat.go
package mat

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

type InitData struct {
	Mode string `json:"mode"`
}

type EquivalenceRequest struct {
	MainPrefixes    string `json:"main_prefixes"`
	NonMainPrefixes string `json:"non_main_prefixes"`
	Suffixes        string `json:"suffixes"`
	Table           string `json:"table"`
}

type EquivalenceResponse struct {
	Response string `json:"response"`
	Type     *bool  `json:"type"`
}

func Init(mode string) error {
	url := "http://localhost:8080/generate"

	data := InitData{
		Mode: mode,
	}
	jsonData, err := json.Marshal(data)
	if err != nil {
		return fmt.Errorf("ошибка при маршаллинге данных: %w", err)
	}

	resp, err := http.Post(url, "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		return fmt.Errorf("ошибка при отправке запроса: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("ошибка: статус ответа %s", resp.Status)
	}

	fmt.Println("Соединение установлено и запрос на /generate выполнен успешно")
	return nil
}

func MembershipQuery(query string) (int, error) {
	url := "http://localhost:8095/checkWord"
	if query == "" {
		query = "ε"
	}

	data := map[string]string{
		"word": query,
	}
	jsonData, err := json.Marshal(data)
	if err != nil {
		fmt.Println("Ошибка при маршаллинге данных:", err)
		return -1, err
	}

	resp, err := http.Post(url, "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		fmt.Println("Ошибка при отправке запроса:", err)
		return -1, err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusOK {
		var response map[string]bool
		err = json.NewDecoder(resp.Body).Decode(&response)
		if err != nil {
			fmt.Println("Ошибка при декодировании ответа:", err)
			return -1, err
		}

		if response["response"] {
			return 1, nil
		} else {
			return 0, nil
		}
	}
	return -1, err
}

func EquivalenceQuery(mainPrefixes, nonMainPrefixes, suffixes, table string) (bool, string, error) {
	url := "http://localhost:8095/checkTable"

	requestData := EquivalenceRequest{
		MainPrefixes:    mainPrefixes,
		NonMainPrefixes: nonMainPrefixes,
		Suffixes:        suffixes,
		Table:           table,
	}

	jsonData, err := json.Marshal(requestData)
	if err != nil {
		return false, "", fmt.Errorf("ошибка при маршаллинге данных: %w", err)
	}

	resp, err := http.Post(url, "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		return false, "", fmt.Errorf("ошибка при отправке запроса: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return false, "", fmt.Errorf("ошибка: статус ответа %s", resp.Status)
	}

	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return false, "", fmt.Errorf("ошибка при чтении тела ответа: %w", err)
	}
	//fmt.Printf("Тело HTTP-ответа: %s\n", string(bodyBytes))

	var response EquivalenceResponse
	err = json.Unmarshal(bodyBytes, &response)
	if err != nil {
		return false, "", fmt.Errorf("ошибка при декодировании ответа: %w", err)
	}
	if response.Type == nil {
		return true, "", nil
	} else {
		return false, response.Response, nil
	}
}

File: lab2/LearnerRefal_Rust/src/mat.rs
use reqwest::blocking::{Client, Response};
use serde::{Deserialize, Serialize};
use std::error::Error;
use std::fmt;

#[derive(Serialize)]
struct InitData {
    mode: String,
}

#[derive(Serialize)]
struct MembershipData {
    word: String,
}

#[derive(Serialize)]
struct EquivalenceRequest {
    main_prefixes: String,
    non_main_prefixes: String,
    suffixes: String,
    table: String,
}

#[derive(Deserialize)]
struct EquivalenceResponse {
    response: Option<String>,
    #[serde(rename = "type")]
    response_type: Option<bool>,
}

#[derive(Debug)]
struct HttpResponseError {
    message: String,
}

impl fmt::Display for HttpResponseError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl Error for HttpResponseError {}

impl HttpResponseError {
    fn new(message: &str) -> Self {
        HttpResponseError {
            message: message.to_string(),
        }
    }
}

pub fn init(mode: &str) -> Result<(), Box<dyn Error>> {
    let client = Client::new();
    let url = "http://localhost:8080/generate";
    let data = InitData {
        mode: mode.to_string(),
    };

    let resp: Response = client.post(url).json(&data).send()?;
    
    if resp.status().is_success() {
        println!("Соединение установлено и запрос на /generate выполнен успешно");
        Ok(())
    } else {
        Err(Box::new(HttpResponseError::new(&format!(
            "Ошибка: статус ответа {}",
            resp.status()
        ))))
    }
}

pub fn membership_query(query: &str) -> Result<i32, Box<dyn Error>> {
    let client = Client::new();
    let url = "http://localhost:8095/checkWord";
    let word = if query.is_empty() { "ε".to_string() } else { query.to_string() };
    let data = MembershipData { word };

    let resp: Response = client.post(url).json(&data).send()?;
    
    if resp.status().is_success() {
        let response: serde_json::Value = resp.json()?;
        
        if response["response"].as_bool().unwrap_or(false) {
            Ok(1)
        } else {
            Ok(0)
        }
    } else {
        Err(Box::new(HttpResponseError::new(&format!(
            "Ошибка: статус ответа {}",
            resp.status()
        ))))
    }
}

pub fn equivalence_query(
    main_prefixes: &str,
    non_main_prefixes: &str,
    suffixes: &str,
    table: &str,
) -> Result<(bool, String), Box<dyn Error>> {
    let client = Client::new();
    let url = "http://localhost:8095/checkTable";
    let request_data = EquivalenceRequest {
        main_prefixes: main_prefixes.to_string(),
        non_main_prefixes: non_main_prefixes.to_string(),
        suffixes: suffixes.to_string(),
        table: table.to_string(),
    };

    let resp: Response = client.post(url).json(&request_data).send()?;
    
    if resp.status().is_success() {
        let response_body: EquivalenceResponse = resp.json()?;
        
        if let Some(response_type) = response_body.response_type {
            Ok((false, response_body.response.unwrap_or_default()))
        } else {
            Ok((true, String::new()))
        }
    } else {
        Err(Box::new(HttpResponseError::new(&format!(
            "Ошибка: статус ответа {}",
            resp.status()
        ))))
    }
}


